repo/repo.go

```diff
@@ -71,62 +71,126 @@ func FindExternalRepo(repoRoot, name string) (string, error) {
 func ListRepositories(workspace *rule.File) (repos []*rule.Rule, repoFileMap map[string]*rule.File, err error) {
 	repoIndexMap := make(map[string]int)
 	repoFileMap = make(map[string]*rule.File)
+	visited := make(map[string]bool)
 	for _, repo := range workspace.Rules {
 		if name := repo.Name(); name != "" {
 			repos = append(repos, repo)
 			repoFileMap[name] = workspace
 			repoIndexMap[name] = len(repos) - 1
 		}
 	}
-	extraRepos, err := parseRepositoryDirectives(workspace.Directives)
-	if err != nil {
+	if err := loadExtraRepos(workspace, &repos, repoFileMap, repoIndexMap); err != nil {
 		return nil, nil, err
 	}
-	for _, repo := range extraRepos {
-		if i, ok := repoIndexMap[repo.Name()]; ok {
-			repos[i] = repo
-		} else {
-			repos = append(repos, repo)
-		}
-		repoFileMap[repo.Name()] = workspace
-	}
 
 	for _, d := range workspace.Directives {
 		switch d.Key {
 		case "repository_macro":
-			f, defName, err := parseRepositoryMacroDirective(d.Value)
+			f, defName, leveled, err := ParseRepositoryMacroDirective(d.Value)
 			if err != nil {
 				return nil, nil, err
 			}
-			f = filepath.Join(filepath.Dir(workspace.Path), filepath.Clean(f))
-			macroFile, err := rule.LoadMacroFile(f, "", defName)
-			if err != nil {
-				return nil, nil, err
-			}
-			for _, repo := range macroFile.Rules {
-				if name := repo.Name(); name != "" {
-					repos = append(repos, repo)
-					repoFileMap[name] = macroFile
-					repoIndexMap[name] = len(repos) - 1
-				}
+			repoRoot := filepath.Dir(workspace.Path)
+			f = filepath.Join(repoRoot, filepath.Clean(f))
+			visited[f+"%"+defName] = true
+
+			la := &loadArgs{
+				repoRoot: repoRoot,
+				repos: repos,
+				repoFileMap: repoFileMap,
+				repoIndexMap: repoIndexMap,
+				visited: visited,
 			}
-			extraRepos, err = parseRepositoryDirectives(macroFile.Directives)
-			if err != nil {
+
+			if err := loadRepositoriesFromMacro(la, leveled, f, defName); err != nil {
 				return nil, nil, err
 			}
-			for _, repo := range extraRepos {
-				if i, ok := repoIndexMap[repo.Name()]; ok {
-					repos[i] = repo
-				} else {
-					repos = append(repos, repo)
-				}
-				repoFileMap[repo.Name()] = macroFile
-			}
+			repos = la.repos
```

Comment:
Hmmm, I think it might be a little clearer if we update `loadRepositoriesFromMacro` to return just the `repos` that it sees. Then we can remove `repos` from `loadArgs` and from here we can call:
```
repos = append(repos, loadedRepos...)

Reply:
We are unable to do this because `len(repos)` is required to update `repoIndexMap`, so therefore a complete version of repos must be passed every time we are gathering new repos for `repoFileMap`. This is so a repo can easily be switched without needing to search through the `repos`. Used here:
```
	for _, repo := range extraRepos {
		if i, ok := repoIndexMap[repo.Name()]; ok {
			(*repos)[i] = repo
		} else {
			*repos = append(*repos, repo)
		}
		repoFileMap[repo.Name()] = f
	}
```